### PR TITLE
Make SpanlessEq more consistent

### LINKED ISSUE #10267: Make `SpanlessEq` more consistent
### Description

The `SpanlessEq::eq_expr` implementation has a wildcard match arm that just returns `false`. 

https://github.com/rust-lang/rust-clippy/blob/d227f18c2eb7c1d21d590ea4d8708d8dc5d01308/clippy_utils/src/hir_utils.rs#L305

This is inconsistent with `SpanlessHash`, which implements hashing for each expressions type. 

https://github.com/rust-lang/rust-clippy/blob/d227f18c2eb7c1d21d590ea4d8708d8dc5d01308/clippy_utils/src/hir_utils.rs#L785-L789

`PartialEq` and `Hash` implementations have to be consistent. The same should apply to `SpanlessEq` and `SpanlessHash`

### Version

```text
rustc 1.69.0-nightly (d7948c843 2023-01-26)
binary: rustc
commit-hash: d7948c843de94245c794e8c63dd4301a78bb5ba3
commit-date: 2023-01-26
host: x86_64-unknown-linux-gnu
release: 1.69.0-nightly
LLVM version: 15.0.7
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by gernot-ohner**
1) Remove wildcard as requested in https://github.com/rust-lang/rust-clippy/issues/10267.
2) Implement `hir_utils::eq_expr` for `ExprKind::Closure`, `ExprKind::ConstBlock`, `ExprKind::InlineAsm` and `ExprKind::Yield`.
3) Reorder branches of `hir_utils::eq_expr` to be in alphabetical order.

---
changelog: none

### CODE REVIEW

File: clippy_utils/src/hir_utils.rs [hunk 1]
```diff
@@ -329,24 +333,70 @@ impl HirEqInterExpr<'_, '_, '_> {
...
             },
             (ExprKind::Ret(l), ExprKind::Ret(r)) => both(l, r, |l, r| self.eq_expr(l, r)),
-            (ExprKind::Path(l), ExprKind::Path(r)) => self.eq_qpath(l, r),
             (&ExprKind::Struct(l_path, lf, ref lo), &ExprKind::Struct(r_path, rf, ref ro)) => {
                 self.eq_qpath(l_path, r_path)
                     && both(lo, ro, |l, r| self.eq_expr(l, r))
                     && over(lf, rf, |l, r| self.eq_expr_field(l, r))
             },
             (&ExprKind::Tup(l_tup), &ExprKind::Tup(r_tup)) => self.eq_exprs(l_tup, r_tup),
+            (&ExprKind::Type(le, lt), &ExprKind::Type(re, rt)) => self.eq_expr(le, re) && self.eq_ty(lt, rt),
```
Comment (flip1995) [inline clippy_utils/src/hir_utils.rs:350]: Why doesn't this trigger the `match_same_arms` lint?

Comment (Jarcho) [reply clippy_utils/src/hir_utils.rs:350]: `SpanlessHash` doesn't treat `{ x }` and `x` as the same.

**Comment by gernot-ohner** [comment]
Thanks for the review!

I wasn't sure whether the implementation of the branch for `Yield` was correct in ignoring comparing the expression and ignoring the `YieldSource`. My impression is that that's correct, but I'd defer to your opinion here. 

I'm also unsure what the preferred procedure is: Should I create another PR for the changes in `SpanlessHash` or it in this one?

**Comment by dswij** [comment]
> I'm also unsure what the preferred procedure is: Should I create another PR for the changes in SpanlessHash or it in this one?

If the scope isn't too big, and it's (kinda) related like this one, I'd say it's fine to do it in one PR.

But doing it in separate PRs is not really a problem as well

**Comment by gernot-ohner** [comment]
In that case, I'd like to do it in a separate PR. 
Is this PR ready to be merged then? Or would you like to have anything else done here?

**Comment by flip1995** [comment]
> Should I create another PR for the changes in `SpanlessHash` or it in this one?

As the implementations of `Hash` an `Eq` should always agree ([`std::Hash`](https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq)), I'd prefer to do this as an additional commit in this PR.

**Review by flip1995 — APPROVED** [review]
Just noticed that `hash_expr` already was complete. Thanks for doing the plumbing work of sorting it anyway.

**Pull request merged**